### ISSUE TITLE
feature request: systemctl could provide user more information when a dependency fails

### ISSUE THREAD

**Issue by archenemies** [systemd/systemd#793]
I'm trying to implement some functionality which will be run before my system hibernates. In debugging this, I get messages like:

```
$ sudo systemctl hibernate
A dependency job for hibernate.target failed. See 'journalctl -xe' for details.
```

Running 'journalctl -xe' shows a lot of lines, some with timestamps that make them irrelevant, some red and some white (I'm not sure why some are red). Running 'systemctl status' is not helpful:

```
$ sudo systemctl status hibernate.target
● hibernate.target - Hibernate
   Loaded: loaded (/usr/lib/systemd/system/hibernate.target; static; vendor preset: disabled)
      Active: inactive (dead)
           Docs: man:systemd.special(7)

Jul 30 15:34:04 amnon systemd[1]: Dependency failed for Hibernate.
Jul 30 15:34:04 amnon systemd[1]: hibernate.target: Job hibernate.target/start failed with result 'dependency'.
```

I would have thought that with all of the meta-information present in the systemd configuration, there would be a way for systemd to report why something failed, with more specifics than just "a dependency failed".

More about what I'm trying to do (and I would also appreciate help debugging it!) - I have a unit file root-hibernate.service:

```
[Unit]
Description=Local system hibernate actions
Before=hibernate.target

[Service]
Type=simple
ExecStart=/etc/rc-hibernate.local

[Install]
RequiredBy=hibernate.target
```

After 'systemctl hibernate', 'journalctl -xe' shows:

```
Jul 30 15:34:02 amnon systemd[1]: root-hibernate.service: Main process exited, code=exited, status=1/FAILURE
Jul 30 15:34:02 amnon systemd[1]: root-hibernate.service: Unit entered failed state.
Jul 30 15:34:02 amnon systemd[1]: root-hibernate.service: Failed with result 'exit-code'.
Jul 30 15:34:04 amnon kernel: PM: Syncing filesystems ... done.
Jul 30 15:34:04 amnon kernel: Freezing user space processes ... (elapsed 0.001 seconds) done.
Jul 30 15:34:04 amnon kernel: PM: Marking nosave pages: [mem 0x0009e000-0x000fffff]
...
Jul 30 15:34:04 amnon kernel: PM: thaw of devices complete after 372.652 msecs
Jul 30 15:34:04 amnon kernel: PM: writing image.
Jul 30 15:34:04 amnon kernel: PM: Cannot find swap device, try swapon -a.
Jul 30 15:34:04 amnon kernel: PM: Cannot get swap writer
```

So it looks like even though my service failed (exit 1), the system is still trying to hibernate and only gives up when it finds there is no swap device... However, there are other red lines:

```
Jul 30 15:34:04 amnon kernel: pciehp 0000:00:1c.1:pcie04: Device 0000:03:00.0 already exists at 0000:03:00, cannot hot-add
```

and other lines that could indicate trouble:

```
Jul 30 15:34:04 amnon dhcpcd[445]: en0: carrier lost
```

and in summary the debugging situation is rather confusing to me, as a new user. Thanks!

**Comment by arvidjaar** [comment]
> So it looks like even though my service failed (exit 1), the system is still trying to hibernate 

Yes, that may be confusing. hibernate.target itself does not really do anything - it simply pulls in systemd-hibernate.service which does real job. Now your service runs concurrently with systemd-hibernate.service. So when your service fails it does result in marking hibernate.target as "failed" (due to Requires dependency) but it does not really affect systemd-hibernate.service.

It is usually recommended to hook your own services int sleep.target; this way they do not race with systemd-hibernate (or other similar) services. Also because systemd-hibernate.service Requires sleep.target, failure of your service would abort hibernation attempt.

**Comment by archenemies** [comment]
@arvidjaar: Thank you! Now, given what you said, I tried changing "Before=hibernate.target" to "Before=systemd-hibernate.service" in my unit file, but it still seems to be trying to hibernate after my own service fails. Is that normal?

As far as hooking into sleep.target, I guess you mean I should put "RequiredBy=sleep.target" in the unit file? And "Before=..."? What if I want to run something only upon hibernation, and not on suspend? Thanks!

**Comment by arvidjaar** [comment]
> Is that normal?

Yes. Before/After do not imply RequiredBy/Requires. Your service now runs before sysemd-hibernate.service but failure of your service does not prevent its execution as happened before.

> I guess you mean I should put "RequiredBy=sleep.target" in the unit file?

Yes.

> And "Before=..."?

This is implied in case of Requires in target definition. But won't hurt.

> What if I want to run something only upon hibernation, and not on suspend? 

Good question. I guess it depends on what exactly are you trying to do. Using /usr/lib/systemd/system-sleep hooks may be alternative.

**Comment by archenemies** [comment]
> @arvidjaar /usr/lib/systemd/system-sleep hooks

Thanks! I should have known about those. I was also considering creating an override file for systemd-hibernate.service.

By the way it turns out I wasn't able to hibernate because my swap device didn't match the resume= setting in /proc/cmdline (see "Cannot find swap device" above). Strangely, according to experiments I made today, the kernel lets you hibernate to any device when there is no resume= argument, but when it is there, the swap device has to match the resume device. Not a problem people usually encounter, but I'm building a sytem on a USB stick which I want to work on multiple laptops. So I'll have a pre-shutdown hook which removes the resume= argument from syslinux.cfg (in case of poweroff), and a pre-hibernate hook which puts it back, with the local hardware's swap UUID... Probably more trouble than it's worth, but now I want to finish what I started.

**Comment by archenemies** [comment]
This is off-topic to the reported issue, but it works now, except for a strange D-Bus error (systemd version 222):
$ sudo systemctl hibernate
[--hibernates, then boots and resumes--]
Warning! D-Bus connection terminated.
Failed to wait for response: Connection reset by peer

Now, perhaps relatedly, I don't see logins on new VTs, and "ssh localhost" takes 25 seconds, and:

```
Aug 01 00:20:17 amnon systemd[1]: Caught <BUS>, dumped core as pid 2058.
```

I wanted to point out something that is more on-topic, which is that when I navigate to a virtual terminal so that I can see kernel warnings interspersed with my commands on the console, I get a much better indication of what's going on with hibernate errors. For example, try "swapoff -a" and "systemctl hibernate" on a VT. After the screen comes back, I think there should just be a few messages like "PM: Cannot find swap device" between the command and the new prompt. Whereas running the same commands in X and sorting through "journalctl -xe" I see hundreds of less relevant messages.

**Comment by macropin** [comment]
Can we please get this on track as a feature request? I'm currently raging against the vagarity of the following:

```
localhost ~ # systemctl status docker.service
● docker.service - Docker Application Container Engine
   Loaded: loaded (/etc/systemd/system/docker.service; enabled; vendor preset: disabled)
   Active: inactive (dead)
     Docs: http://docs.docker.com

Feb 05 01:13:18 localhost systemd[1]: Dependency failed for Docker Application Container Engine.
Feb 05 01:13:18 localhost systemd[1]: docker.service: Job docker.service/start failed with result 'dependency'.
```

Is it too much to ask for "Job docker.service/start failed because dependency **foo.service**"
